### ISSUE TITLE
Update WebIDL bindings toolchain

### ISSUE THREAD

**Issue by wingo** [emscripten-core/emscripten#8476]
The WebIDL bindings generator is out of date relative to current WebIDL.  Current WebIDL differs significantly from the dialect implemented by emscripten.

Additionally the current WebIDL bindings generator is not very expressive.  For example it doesn't understand `std::vector` or `std::shared_ptr`.  It would be nice to update it.

See also https://github.com/wingo/pictie/issues/2 for stumbling blocks that I found when attempting to port from embind to webidl.

I can take this one within the next month or so, fwiw.

**Comment by kripken** [comment]
@wingo That would be great! :)

**Comment by jgravelle-google** [comment]
It's almost-surely worth keeping the wasm WebIDL Bindings proposal in mind when updating Emscripten's WebIDL bindings generator: https://github.com/WebAssembly/webidl-bindings
Which is to say, the more compatible the bindings can be, the better. In particular, how do we want to represent these things in C++? Also worth figuring out where the webidl-bindings proposal is missing something that we want to bind, and whether we can have that inform the spec. Ditto for things that are technically supported, but cumbersome or inefficient. In principle we should be able to implement this in terms of a polyfill, though that might need backend or binaryen work to support generating the custom section.

This is something I've been planning on getting to in the next 2-3 months, but I am pretty okay with outsourcing it :D. Let me know when you get started, I have a number of thoughts about this area that would probably be worth crystallizing into an actual discussion at (or before) that point.

**Comment by wingo** [comment]
Thanks for the note, @jgravelle-google !  And thanks for the link to the webidl-bindings spec, I hadn't read that yet, and it does bring up a number of thorny issues :)  I haven't really considered them so here's some unstructured thoughts.  Would be glad to have a call soonish; we can coordinate over email if you like (wingo at igalia.com).

To simplify the discussion, let's call facilities defined in C++-compiled-to-WebAssembly as "native" facilities, and facilities that are imported from the embedder via WebIDL as "foreign" facilities.  Perhaps you have better terms; corrections welcome :)

In the current WebIDL bindings generator, we only have native definitions.  No WebIDL type or operation is foreign.

In a world where WebAssembly can import operations and types from the embedder, that will no longer be the case.  We'll at the very least need a way to mark a WebIDL fragment as being "imported" -- that the bindings generator should generate stubs to expose those foreign operations to C++, instead of bindings to export native operations.

I don't fully understand how anyref gets compiled to WebAssembly (is there a magical type for that?), but however it works, foreign interfaces should probably be wrapped as `anyref` on the C++ side.  Native interfaces of course don't need wrappers.  Not sure how to avoid wrappers in foreign callbacks with native values though.

When exposing native facilities, I think it should be a goal to allow as much freedom of representation as possible to the native code.  Many native facilities will be from outside the web platform, and they might target non-WebAssembly uses as well, so we can't expect native code to (e.g.) switch to a different internal string representation.

Beyond those concerns... I actually don't know what compatibility problems to watch out for.  I am assuming that in the short and medium term:

 * JS will continue to consume WebIDL-described interfaces from the browser
 * With the Wasm webidl-bindings proposal, Wasm will begin to consume WebIDL-described interfaces from the browser, without going through JS
 * With the emscripten WebIDL bindings generator, JS will consume WebIDL-described interfaces from Wasm

But, but the browser won't consume WebIDL-described interfaces from JS or Wasm, and JS itself won't define any WebIDL interfaces.  Perhaps I am wrong here and I certainly haven't thought about it enough.  I am assuming that C++ will know whether an interface is foreign or native, and it is OK to have slight differences in calling conventions (e.g. C++ can directly use native constructors that are not exported to Wasm).  What do you think?

**Comment by wingo** [comment]
Hum, the webidl-bindings proposal is still sinking in :)  If I understand correctly, the webidl-bindings project can be implemented by a new kind of bindings generator that:

 * generates header files for native-ish interfaces corresponding to the imported WebIDL interfaces
 * generates special webidl type and bindings sections (https://github.com/WebAssembly/webidl-bindings/blob/master/proposals/webidl-bindings/Explainer.md#web-idl-types), and somehow links them into the resulting wasm, ideally with good integration with DCE
 * generates js glue code to actually provide bound operations in the `importObject`

And additionally as a polyfill, before browsers support the webidl sections, the bindings generator could also or instead generate C++ files corresponding to the bindings.  AFAIU WebAssembly engine support is only interesting in that it (1) decreases download size and (2) eventually will allow for direct calls to platform WebIDL-described interfaces instead of bouncing through JS.  Is that right?

**Comment by wingo** [comment]
Just an update here as I finally started with this this week.  I extracted the webidl bindings generator to a standalone repo, while I was rebasing:

  https://gitlab.com/wingo/webidl-embindgen/tree/update-webidl-parser

I then extracted the set of patches to WebIDL.py from upstream mozilla since the branch point in 2014.  I applied them all then filtered out the patches that added non-standard Mozilla things (most of these described here: https://developer.mozilla.org/en-US/docs/Mozilla/WebIDL_bindings).

I then went back and added https://github.com/emscripten-core/emscripten/blob/master/tests/webidl/test.idl and the corresponding generated test.cpp and test.idl, and went through and made sure that the binding generator worked at every commit.  Of course at the commit where arrays were removed, things stopped working, so I adapted the bindings generator to work with sequences instead.

(Incidentally, in some contexts we should use FrozenArray<T> rather than sequence<T>; https://heycam.github.io/webidl/#idl-frozen-array.  However Mozilla's parser doesn't support this yet.  This can be fixed up later.)

It turned out that the adaptations to webidl_binder.py can apply to the current parser and bindings generator with no problems.  I will submit these changes as a separate patch (https://gitlab.com/wingo/webidl-embindgen/commit/53bbeedaa50599cc8e10770a600240efb6da4cd9).  However I need to make sure that ammo.js works, and I haven't verified it yet.  When I have that working on Monday I'll submit the patch to adapt the bindings generator and test IDL file, and a patch for ammo.idl to use sequences rather than arrays; then I'll submit a patch series to update the WebIDL parser.

**Comment by kripken** [comment]
Nice, sounds good @wingo !

**Comment by VirtualTim** [comment]
This is cool. I was considering doing something similar myself, but balked at the number of differences.
I took a quick peek at the diff between your version of WebIDL.py and the current Emscripten one.

Seems like the main differences are:
Adding USVString and JSON (non-standard?) types, adding "[IDLPromiseType](https://gitlab.com/wingo/webidl-embindgen/blob/update-webidl-parser/third_party/WebIDL.py#L6215-6216)" (looks interesting, but couldn't figure out how that worked)
Adding [Clamp], [EnforceRange], [LenientSetter] extended attributes,
Added IDLCallbackType (looks like that didn't work before?)
Jsonifiers have been removed (but I can't see if that was used/documented anywhere).

Overall this looks good/interesting, but kind of hard to judge just by looking at 5k lines of code.
Callback and promise types sound good.
